### ISSUE TITLE
Activation Functions

### ISSUE THREAD

**Issue by stefanpeidli** [stefanpeidli/GoNet#10]
From C. Bishop Neuronal Networks for Pattern Recognition, pages: 126ff
For the policy net, it is practical to use the softmax function, because it yields a distribution as output of the NN, which is what the policy net needs.

**So how to choose the activation function(s) of the hidden layers?**
Commonly used and proposed by Bishop are sigmoidal functions like the tanh and the logistic sigmoid (note that those two only differ through a linear trafo).

> The  logistic sigmoid  (4.10)  is often  used  for  the  hidden  units  of a  multi-layer network.  However,  there may  be some small practical  advantage  in using a  'tanh' activation  function [...]. Empirically,  it  is  often found  that  'tanh'  activation  functions  give  rise to  faster  convergence  of  training algorithms  than  logistic  functions.

This justifies our choice of tanh as activation function of the hidden layers.

**Comment by stefanpeidli** [comment]
AlphaGo uses Kernels. Should we too?

**Comment by stefanpeidli** [comment]
For Kernels, if we'd ever want them, I will open a new issue. This issue is only for the activation functions in hidden layers and the output layer.

**Comment by stefanpeidli** [comment]
As Paddy said, one proposal would be using Elu.
https://www.quora.com/How-does-ELU-activation-function-help-convergence-and-whats-its-advantages-over-ReLU-or-sigmoid-or-tanh-function

TODO: try this out

**Comment by paddyooo** [comment]
yes, i believe we need kernels if we want it to become great at go!

**Comment by stefanpeidli** [comment]
Seems like we should experiment with different activation functions than sigmoids.

I'm open for suggestions.

**Comment by paddyooo** [comment]
http://peterroelants.github.io/posts/neural_network_implementation_intermezzo01/
cross entropy apparently works with the logistic function as activation function and cross entropy cost function as error function

**Comment by stefanpeidli** [comment]
The site mentioned above considers a 2-classification problem. We do not.
This might be interesting though when creating a ValueNet (since this is basically a 3-classification problem).

**Comment by stefanpeidli** [comment]
For the moment, results strongly speak for keeping sigmoids as activation functions. Until we have a reason to revisit ReLus and similar, this issue has rather subordinate role and thus is closed.
